velox/exec/WindowPartition.cpp

```diff
@@ -263,20 +263,22 @@ void WindowPartition::updateKRangeFrameBounds(
     const vector_size_t* rawPeerBounds,
     vector_size_t* rawFrameBounds) const {
   column_index_t orderByColumn = sortKeyInfo_[0].first;
-  RowColumn frameRowColumn = columns_[frameColumn];
+  column_index_t mappedFrameColumn = inputMapping_[frameColumn];
 
   vector_size_t start = 0;
   vector_size_t end;
   for (auto i = 0; i < numRows; i++) {
     auto currentRow = startRow + i;
-    bool frameIsNull = RowContainer::isNullAt(
-        partition_[currentRow],
-        frameRowColumn.nullByte(),
-        frameRowColumn.nullMask());
 
-    // For NULL values, CURRENT ROW semantics apply. So get frame bound from
-    // peer buffer.
-    if (frameIsNull) {
+    // If the frame is NULL or 0 preceding or 0 following then the current row
+    // has same values for order by and frame column. In that case
+    // the bound matches the peer row for this row.
+    if (data_->compare(
+            partition_[currentRow],
+            partition_[currentRow],
+            orderByColumn,
+            mappedFrameColumn,
+            flags) == 0) {
```

Comment:
The original code sets the current frame to the peer rows when the frame column is null at this row; the new code sets the current frame to peer rows when both order-by column and frame column have the same value, e.g., both are nulls. Is it ever possible that the order-by column is not null but the frame column is null? (In this situation, the old code and new code behave differently?)

Reply:
@kagamiori : If order-by column is NULL, then frame column (always derived by doing k +/- order by column) is always NULL as well. In the old code, I was skipping the equality check based on this assumption. 

After this fix I realized we have to do the equals check for every row now, so then the NULL situation gets caught by it as well. So removed the old NULL handling.

Reply:
> If order-by column is NULL, then frame column (always derived by doing k +/- order by column) is always NULL as well.

Got it. Does the derivation of frame column happen inside Velox or in the customer code (e.g., Prestissimo code)? Could we add a comment about this rule?

Reply:
@kagamiori : The computation happens in the service (Prestissimo and Spark code).

When I think about this a bit more...The frame column should be NULL when order by column is NULL, but a frame column could be NULL even when the order by column is not NULL e.g. If using a column value for the range (instead of constant), it could contain NULL causing frame column to be NULL. In this case the old code was incorrectly returning the current peer, while now it will look for NULL In the order by column which is the correct behavior.

I'll work on another test.

Reply:
@kagamiori : I verified the condition above with Presto and seems like it errors constructing the offset column in such a case

```
presto:tpch> select a, sum(a) over (partition by a order by b range between c preceding and unbounded following) from (values (1, null, 1), (1, 2, null), (1, 3, 3)) as t(a, b, c);

Query 20240621_190434_00002_jbv3j failed: Window frame offset value must not be negative or null
```

I'll add a comment and check to the code.

Reply:
Done.